src/System.Management.Automation/namespaces/FileSystemProvider.cs

```diff
@@ -1080,6 +1080,13 @@ protected override bool IsValidPath(string path)
                 }
             }
 
+            // .NET introduced a change where invalid characters are accepted https://learn.microsoft.com/en-us/dotnet/core/compatibility/2.1#path-apis-dont-throw-an-exception-for-invalid-characters
+            // We need to check for invalid characters ourselves
+            if (Path.GetDirectoryName(path).IndexOfAny(Path.GetInvalidPathChars()) != -1 || Path.GetFileName(path).IndexOfAny(Path.GetInvalidFileNameChars()) != -1)
```

Comment:
I think this logic might not be incorrect, on non-Windows you get the following values for these functions

+ `GetInvalidPathChars()` - `0x00 NULL`
+ `GetInvalidFileNameChars()` - Above plus `0x2F '/'`

On Windows these are

+ `GetInvalidPathChars()` - `0x00` to `0x1F`, `0x7C '|'`
+ `GetInvalidFileNameChars()` - Above plus `0x22 - '"'`, `0x2A - '*'`, `0x2F - '/'`, `0x3A - ':'`, `0x3C - '<'`, `0x3E - '>'`, `0x3F - '?'`, `0x5C - '\'`

Currently the code here is only going to check those extra chars for the filename but each folder component in the path cannot be one of the invalid FileNameChars. For example `C:\Windows\Foo*Bar\test` is invalid as `Foo*Bar` contains `*` which is an invalid filename but the code here is only going to check `test` with those extra chars.

Even splitting the path may not be enough as you'll need to validate the drive root is a valid path.

Finally NTFS ADS parts in the filename will trigger this check

```powershell
[System.IO.Path]::GetFileName('C:\temp\test\file.txt:$data')
file.txt:$data
```

The ADS portion is separated by `:` but `GetFileName` won't split that out and `:` in there is going to trigger the check.

Reply:
By the time this code hits, the NTFS ADS parts are already removed (along with some other cleanup/normailzation).  There's also checks prior to this on being a valid rooted path.

Reply:
Fair enough, I think my point about checking each path component still stands though

```powershell
PS C:\Users\vagrant-domain> [System.IO.FileInfo]::new("C:\temp\foo*bar\test.txt")

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----


PS C:\Users\vagrant-domain> [System.IO.FileInfo]::new("C:\temp\foobar\test*.txt")

Mode                 LastWriteTime         Length Name
----                 -------------         ------ ----
larhs            1/1/1601 12:00 AM
```

Both are invalid but the changes here will only pick up `test*.txt` as invalid but ignore `foo*bar`.

Reply:
@jborean93 yes, I think checking each path component will be necessary.  Will update PR.